quixstreams/core/stream/stream.py

```diff
@@ -202,53 +206,194 @@ def add_transform(
+            # the easiest check that catches most issues: the "inner" (filtering) sdf
+            # should use same ref as the one being filtered; i.e. sdf[sdf.apply()].
+            if diff_origin != self:
+                raise InvalidOperation(
+                    "SDF's used as filters must originate from the filtered SDF; "
+                    "ex: `sdf[sdf.apply()]`, NOT `sdf[other_sdf.apply()]"
+                )
+
+            # With splitting there are some edge cases where the origin is the same,
+            # but there are various side effects that can occur that we want to avoid.
+            other_path_start = other.tree_root_path(allow_splits=False)[0]
+            if other_path_start.parent != diff_origin:
+                # There is a split at the filtering SDF; still potentially valid
+                if self.tree_root_path(allow_splits=False)[0] != other_path_start:
+                    # This split is not shared by the filtered sdf
+                    raise InvalidOperation(
+                        "SDF's used as filters must originate from the filtered SDF; "
+                        "ex: `sdf[sdf.apply()]`, NOT `sdf[other_sdf.apply()]"
+                    )
+
+        if prune:
+            self.prune(diff[0])
+
         parent = None
-        head = None
         for node in diff:
             # Copy the node to ensure we don't alter the previously created Nodes
             node = copy.deepcopy(node)
             node.parent = parent
             parent = node
-            head = node
-        return head
+        return parent
 
-    def tree(self) -> List[Self]:
+    def tree_root_path(self, allow_splits=True) -> List[Self]:
         """
         Return a list of all parent Streams including the node itself.
 
+        Can optionally stop at a first encountered split with allow_splits=False
+
         The tree is ordered from parent to child (current node comes last).
         :return: a list of `Stream` objects
         """
 
-        tree_ = [self]
         node = self
-        while node.parent:
-            tree_.append(node.parent)
+        tree_ = [node]
+        while (parent := node.parent) and (allow_splits or len(parent.children) < 2):
+            tree_.append(parent)
             node = node.parent
 
         # Reverse to get expected ordering.
         tree_.reverse()
 
         return tree_
 
+    def prune(self, other: Self):
+        """
+        Removes a stream node by looking for where the "other" node is a child within
+        the current and removing it.
+
+        Note this means "other" must share a direct split point with this Stream.
+        :param other: another Stream
+        :return:
+        """
+        node = self
+        while node:
+            if other in node.children:
+                node.children.remove(other)
+                return
+            node = node.parent
+        raise InvalidOperation("Cannot filter with the same SDF more than once.")
+
+    def tree_all_nodes(
```

Comment:
I want to suggest another names here too:

- `tree_all_nodes` -> `full_tree`
- `tree_root_path` -> `root_path`

These two now have different meaning, and with this new feature only "full_tree" is actually a tree, when "root_path" is a section in the tree.